sig-and-wg/sig-core/proposals/ui-api-layer-modularization.md

```diff
@@ -0,0 +1,88 @@
+# UI API Layer Modularization
+
+Created on 2018-29-11 by Paweł Kosiec (@pkosiec).
+
+## Status
+
+Proposed on 2018-11-29.
+
+## Motivation
+
+UI API Layer exposes GraphQL API for the Kyma Console. It consists of resolver logic for different domains, like Kubernetes, Service Catalog, Remote Environments, Kubeless, etc. As Kyma needs to be modularized, users should be able to install only the components they need.
+
+Current approach we use in UI API Layer makes it impossible. During the GraphQL server start, we try to synchronize Stores of all Informers. In a result, if a resource of a specific type doesn't exist, UI API Layer won't start properly, because its Informer will return an error.
+
+## Goal
+
+1. There should be an ability to turn on and off specific UI API Layer modules, even on server runtime.
+1. If user tries to access queries/mutations/subscriptions of a specific domain, which is related with not installed Kyma component, UI API Layer should return an error about uninstalled domain.
+
+## Suggested solution
+
+The solution is based on a Custom Resource, which is defined for all Kyma components which have equivalents in UI API Layer as different domains (`domain` package). In this proposal it is called as `ConsoleBackendModule`.
+
+Every domain resolver object has to implement `PluggableModule` interface. By default, during start, UI API Layer has all domain resolver disabled, which means there are no informers synced for all resources related to GraphQL queries/mutations/subscriptions.
+
+Using generated Informer API for the `ConsoleBackendModule` resource, an event handler is created for every domain resolver. On create/update/delete events the handler is triggered, and it enables/disables according domain resolver.
+
+Enabling a domain resolver means everything what it is currently done in domain resolver constructor (services, informers setup etc), as well as starting informers and waiting for cache synchronization.
+
+Disabling a specific domain resolver stops the informer synchronization, unregisters subscription listeners and replaces resolver logic to generated one, which always returns an error (e.g. `Module {MODULE_NAME} is not installed`). For all `Container` objects, which contain domain resolvers and exported cross-domain dependencies, all exported fields need to be replaced with generated implementation.
+
+Generator could be based on [Impl](https://github.com/josharian/impl) - but it will need a modification: instead of panicking, it should return error gracefully for every method.
```

Comment:
```suggestion
The Generator could be based on [Impl](https://github.com/josharian/impl) but it will need a modification: instead of panicking, it should return an error gracefully for every method.
```